tests/test_markdown_plugin.py

```diff
@@ -0,0 +1,257 @@
+from __future__ import annotations
+
+import itertools
+from contextlib import nullcontext as does_not_raise
+from typing import NamedTuple
+
+import pytest
+
+from porcupine.plugins import markdown
+
+
+class ListItemCase(NamedTuple):
+    id: str
+    line: str
+    expected: bool
+    marks: list[pytest.MarkDecorator] = []
+    raises: Exception | None = None
+
+
+IS_LIST_ITEM_CASES = [
+    ListItemCase(id="# with no separator", line="# item 1", expected=False),
+    ListItemCase(id="# bad separator |", line="#| item 1", expected=False),
+    ListItemCase(id="# bad separator /", line="#/ item 1", expected=False),
+    ListItemCase(id="# bad separator \\", line="#\\ item 1", expected=False),
+    ListItemCase(id="ol bad separator |", line="8| item 1", expected=False),
+    ListItemCase(id="ol bad separator /", line="8/ item 1", expected=False),
+    ListItemCase(id="ol bad separator \\", line="8\\ item 1", expected=False),
+    ListItemCase(id="not a list 1", line="item 1", expected=False),
+    ListItemCase(id="not a list 2", line="   item 1", expected=False),
+    ListItemCase(id="not a list 3", line="          item 1", expected=False),
+    ListItemCase(id="not a list 4", line="& item 1", expected=False),
+    ListItemCase(id="not a list 5", line="^ item 1", expected=False),
+    ListItemCase(id="duplicate token 1", line="-- item 1", expected=False),
+    ListItemCase(id="duplicate token 2", line="--- item 1", expected=False),
+    ListItemCase(id="duplicate token 3", line="- - - item 1", expected=True),
+    ListItemCase(id="duplicate token 4", line="  - item -- 1 -", expected=True),
+    ListItemCase(id="duplicate token 5", line="  -#) item -- 1 -", expected=False),
+    ListItemCase(id="duplicate token 6", line="  *-#)1. item -- 1 -", expected=False),
+]
+
+# test `#` and 0 to 99 numbered lists
+# tests ol with `.` and `)`
+IS_LIST_ITEM_CASES.extend(
+    [
+        ListItemCase(id=f"numbered {i}", line=f"{i}{sep} item 1", expected=True)
+        for i, sep in itertools.product(itertools.chain(range(100), "#"), (".", ")"))
```

Comment:
These are maybe a bit too thorough :) We really don't need more than a thousand green dots when running pytest because of one regex. But more importantly, these two lines aren't very readable. It took me a while to figure out what they do.

Maybe instead add a few specific cases, so that it's easier to understand?

```
ListItemCase(id=f"numbered 1", line=f"1. item 1", expected=True)
ListItemCase(id=f"numbered 1", line=f"1) item 1", expected=True)
ListItemCase(id=f"numbered 4", line=f"4. item 1", expected=True)
ListItemCase(id=f"numbered 4", line=f"4) item 1", expected=True)
ListItemCase(id=f"numbered 42", line=f"42. item 1", expected=True)
ListItemCase(id=f"numbered 42", line=f"42) item 1", expected=True)
ListItemCase(id=f"numbered #", line=f"#. item 1", expected=True)
ListItemCase(id=f"numbered #", line=f"#) item 1", expected=True)
```

Also applies to other `.extend()`s.

Reply:
> We really don't need more than a thousand green dots when running pytest because of one regex

But it's so fun :weary: 

> these two lines aren't very readable

That's fair. It felt readable at the time :sweat: 

> Maybe instead add a few specific cases, so that it's easier to understand?

Yeah I can do this.

On a more broad note, how would you feel about adding https://hypothesis.readthedocs.io/en/latest/ to the test suite?

Reply:
I'm not a big fan of `hypothesis`. IMO tests should be as concrete/specific as possible, such as `ListItemCase(id="numbered 42", line="42) item 1", expected=True)`, whereas the code being tested tends to be written very generally. This ensures that at least the test is correct, because it's ideally very simple.

And yes, I know that the point of `hypothesis` is to catch bugs where a specific string that you didn't think of causes a problem. But they tend to get caught without unit tests in Porcupine. For example, the empty `line.splitlines()` bug caused errors when I opened a file.

I can see `hypothesis` being useful in other projects though. For example, maybe you have functions to escape and unescape a string in some way, and you want to know that they work in all corner cases.

Reply:
I would see hypothesis as purely additive, it wouldn't replace the simple `ListItemCase(id="numbered 42", line="42) item 1", expected=True)`, sanity type tests, but it would replace:

```python
IS_LIST_ITEM_CASES.extend(
    [
        ListItemCase(id=f"numbered {i}", line=f"{i}{sep} item 1", expected=True)
        for i, sep in itertools.product(itertools.chain(range(100), "#"), (".", ")"))
```

> For example, the empty line.splitlines() bug caused errors when I opened a file.

I think there is likely many more of these sorts of things, especially as more and more plugins get added ;)

I think hypothesis would likely help a lot with the flaky tests as well. I should probably move this discussion to an issue.